Python/pylifecycle.c

```diff
@@ -1348,6 +1351,20 @@ Py_FinalizeEx(void)
     PyThreadState *tstate = _PyRuntimeState_GetThreadState(runtime);
     PyInterpreterState *interp = tstate->interp;
 
+    // Finalize sub-interpreters.
+    runtime->interpreters.allow_new = 0;
+    PyInterpreterState *curr_interp = PyInterpreterState_Head();
+    PyInterpreterState *next_interp;
+    while (curr_interp != NULL) {
+        next_interp = PyInterpreterState_Next(curr_interp);
+        if (curr_interp != PyInterpreterState_Main()) {
+            PyThreadState_Swap(curr_interp->tstate_head);
+            Py_EndInterpreter(curr_interp->tstate_head);
+        }
+        curr_interp = next_interp;
+    }
+    PyThreadState_Swap(tstate);
```

Comment:
As we discussed, I recommend the following:

* fail if not in the main interpreter
* emit a `ResourceWarning` if there are any lingering interpreters

So something like this:

```suggestion
    if (interp != PyInterpreterState_Main()) {
        # XXX Set an exception (or at least print a message to stderr).
        return -1;
    }
    
    // Finalize sub-interpreters.
    runtime->interpreters.allow_new = 0;
    PyInterpreterState *curr_interp = PyInterpreterState_Head();
    PyInterpreterState *next_interp;
    int64_t num_destroyed = 0;
    while (curr_interp != NULL) {
        next_interp = PyInterpreterState_Next(curr_interp);
        if (curr_interp != PyInterpreterState_Main()) {
            num_destroyed += 1;
            PyThreadState_Swap(curr_interp->tstate_head);
            Py_EndInterpreter(curr_interp->tstate_head);
        }
        curr_interp = next_interp;
    }
    PyThreadState_Swap(tstate);
    if (num_destroyed > 0) {
        if (PyErr_ResourceWarning(..., 1, "extra %zd interpreters", num_destroyed) != 0) {
            ...
        }
    }
```